From 09773875984188e33a95522344f7a64ba55331dc Mon Sep 17 00:00:00 2001
From: Sujeet Baranwal <sbaranwal@marvell.com>
Date: Thu, 12 Dec 2019 12:53:42 -0800
Subject: [PATCH 0427/1921] mmc: octeontx2: speed limit for tx2-c0

Speed is limited to 150MHz for C0 in ddr modes.

Change-Id: I03206fb354be207c4eef8095cc736e99dfb89ae2
Signed-off-by: Sujeet Baranwal <sbaranwal@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/20292
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/23279
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/26922
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/mmc/host/cavium.c |  2 ++
 drivers/mmc/host/cavium.h | 14 ++++++++++++++
 2 files changed, 16 insertions(+)

diff --git a/drivers/mmc/host/cavium.c b/drivers/mmc/host/cavium.c
index 9881bf41c20c..dff5162db483 100644
--- a/drivers/mmc/host/cavium.c
+++ b/drivers/mmc/host/cavium.c
@@ -1376,6 +1376,8 @@ static u32 max_supported_frequency(struct cvm_mmc_host *host)
 		/* Erratum is only applicable pass A0 */
 		if (is_mmc_otx2_A0(host))
 			max_frequency = MHZ_100;
+		else if	(is_mmc_otx2_C0(host))
+			max_frequency = MHZ_150;
 	}
 	return max_frequency;
 }
diff --git a/drivers/mmc/host/cavium.h b/drivers/mmc/host/cavium.h
index f38353171678..ec10ed476128 100644
--- a/drivers/mmc/host/cavium.h
+++ b/drivers/mmc/host/cavium.h
@@ -30,10 +30,15 @@
 #define PCI_SUBSYS_DEVID_9XXX	0xB
 #define PCI_SUBSYS_DEVID_95XX	0xB3
 
+/* Chip revision Id */
+#define REV_ID_0 0
+#define REV_ID_2 2
+
 #define KHZ_400 (400000)
 #define MHZ_26  (26000000)
 #define MHZ_52  (52000000)
 #define MHZ_100 (100000000)
+#define MHZ_150 (150000000)
 #define MHZ_200 (200000000)
 
 /* octtx2: emmc interface io current drive strength */
@@ -329,6 +334,15 @@ static inline bool is_mmc_otx2_A0(struct cvm_mmc_host *host)
 #endif
 }
 
+static inline bool is_mmc_otx2_C0(struct cvm_mmc_host *host)
+{
+	struct pci_dev *pdev = host->pdev;
+	u32 chip_id = (pdev->subsystem_device >> 12) & 0xF;
+
+	return (pdev->revision == REV_ID_2) &&
+		(chip_id == PCI_SUBSYS_DEVID_9XXX);
+}
+
 static inline bool is_mmc_95xx(struct cvm_mmc_host *host)
 {
 	struct pci_dev *pdev = host->pdev;
-- 
2.31.1

